tests/test_utils.py

```diff
@@ -55,3 +56,38 @@
 )
 def test_guess_intype(file_name, first_line, expected):
     assert _guess_intype(file_name, [first_line]) == expected
+
+
+@pytest.mark.parametrize(
+    "input,output",
+    (
+        # Safe URLs are returned unmodified.
+        (
+            {"url": "https://example.com"},
+            {"url": "https://example.com"},
+        ),
+        # Unsafe URLs in the url field are modified, while left untouched on
+        # other fields.
+        (
+            {
+                "a": {"b", "c"},
+                "d": "https://example.com/ a",
+                "url": "https://example.com/ a",
+            },
+            {
+                "a": {"b", "c"},
+                "d": "https://example.com/ a",
+                "url": "https://example.com/%20a",
+            },
+        ),
+        # NOTE: We use w3lib.url.safe_url_string for escaping. Tests covering
+        # the URL escaping logic exist upstream.
```

Comment:
Could you please add a test for an URL with a fragment, just to make sure we keep fragments?